### PR TITLE
NEW: `create_crypto_box` optimisation.

### PR DESCRIPTION
**Pull request by melsomino**
When user creates crypto box, SDK encrypts provided secret information using provided password
  and salt.
  When SDK encrypts secret, it derives encryption key from password and salt using `scrypt`
  function which takes a lot of CPU time (about 1 second).
  So when the user creates many crypto boxes using same password and salt, it is required a lot
  of time (about 12 seconds for 10 crypto boxes).
  With optimisations introduced in this version the SDK caches mapping (password+salt => encryption key)
  for approximately 2 seconds.
  So when the user creates many crypto boxes using same password and salt, it is required only
  a 1.2 seconds for 10 crypto boxes.

### CODE REVIEW

**Review by elasticLove1 — APPROVED** [review]

File: ton_client/src/crypto/boxes/crypto_box/derived_keys.rs [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+struct DerivedKeysCache {
+    keys: Vec<DerivedKey>,
+    env: Arc<ClientEnv>,
+}
+
+impl DerivedKeysCache {
+    fn touch(&mut self, hash: &SecretHash) -> Option<&SecretBuf> {
+        for key in &mut self.keys {
+            if key.hash.0 == hash.0 {
+                println!("use cached key");
```
Comment (AlexeyVavilin) [inline ton_client/src/crypto/boxes/crypto_box/derived_keys.rs:49]: Remove print or replace with trace logging

File: ton_client/src/crypto/boxes/crypto_box/derived_keys.rs [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+            cache: Arc::new(RwLock::new(DerivedKeysCache {
+                keys: Vec::new(),
+                env,
+            })),
+        }
+    }
+
+    pub(crate) fn derive(&self, password: &[u8], salt: &str) -> ClientResult<SecretBuf> {
+        let hash = DerivedKey::calc_hash(password, salt);
+        if let Some(existing) = { self.write_cache().touch(&hash).map(|x| x.clone()) } {
```
Comment (AlexeyVavilin) [inline ton_client/src/crypto/boxes/crypto_box/derived_keys.rs:107]: Are curly braces needed here? Guard is dropped after experssion is executed

File: CHANGELOG.md [hunk 3]
```diff
@@ -4,9 +4,24 @@ All notable changes to this project will be documented in this file.
...
+  With the optimisations introduced in this version the library stores the 
+  pair (password+salt => encryption key) in internal cache for approximately 2 seconds.
+  So when a user creates many crypto boxes at a time using the same password and salt, 
+  library uses cached information to skip heavy calculations. As a result now it takes only 
+  a second to create 10 crypto boxes.  
+
 ### Fixed
 
-- Some enum types wasn't properly presented in api.json (some types that uses serde(content="value"))
+- Some enum types ware not properly presented in api.json (some types that use serde(content="value"))
```
Comment (AlexeyVavilin) [inline CHANGELOG.md:24]: Wrong fix

**Review by AlexeyVavilin — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 3 above)
Comment (elasticLove1) [reply CHANGELOG.md]: I fixed the fix)))

**Pull request closed without merging**